File: ui/store/settings/sections/terminal.ts
import { type SettingsSection } from '@/store/settings/types';

export type TerminalSettings = SettingsSection;

export const initialState: TerminalSettings = {
  id: 'terminal',
  label: 'Terminal',
  description: 'Customize the terminal',
  icon: 'LuTerminalSquare',
  settings: {
    defaultShell: {
      label: 'Default Shell',
      description: 'The default shell to use for new terminals',
      visible: true,
      type: 'text',
      default: '/bin/sh',
      value: '/bin/sh',
    },
    theme: {
      label: 'Theme',
      description: 'Choose a theme for the terminal',
      visible: true,
      type: 'select',
      default: 'default',
      value: 'default',
      options: [
        { value: 'default', label: 'Default' },
        { value: 'dark', label: 'Dark' },
        { value: 'light', label: 'Light' },
      ],
    },
    fontSize: {
      label: 'Font Size',
      description: 'Set the font size for the terminal',
      visible: true,
      type: 'integer',
      default: 14,
      value: 14,
    },
    fontFamily: {
      label: 'Font Family',
      description: 'Set the font family for the terminal',
      visible: true,
      type: 'select',
      default: 'monospace',
      value: 'monospace',
      options: [
        { value: 'monospace', label: 'Monospace' },
        { value: 'sans-serif', label: 'Sans Serif' },
        { value: 'serif', label: 'Serif' },
      ],
    },
    fontWeight: {
      label: 'Font Weight',
      description: 'Set the font weight for the terminal',
      visible: true,
      type: 'select',
      default: 'normal',
      value: 'normal',
      options: [
        { value: 'normal', label: 'Normal' },
        { value: 'bold', label: 'Bold' },
      ],
    },
    cursorStyle: {
      label: 'Cursor Style',
      description: 'Set the cursor style for the terminal',
      visible: true,
      type: 'select',
      default: 'block',
      value: 'block',
      options: [
        { value: 'block', label: 'Block' },
        { value: 'underline', label: 'Underline' },
        { value: 'bar', label: 'Bar' },
      ],
    },
    cursorWidth: {
      label: 'Cursor Width',
      description: 'Set the cursor width for the terminal',
      visible: true,
      type: 'integer',
      default: 2,
      value: 2,
    },
    cursorBlink: {
      label: 'Cursor Blink',
      description: 'Enable or disable cursor blinking for the terminal',
      visible: true,
      type: 'toggle',
      default: true,
      value: true,
    },
  },
};

File: internal/settings/terminal.go
package categories

import (
	"fmt"
	"os"
	"os/exec"

	"github.com/omniviewdev/plugin-sdk/settings"
)

//nolint:gochecknoglobals // we're providing this as a package level variable
var Terminal = settings.Category{
	ID:          "terminal",
	Label:       "Terminal",
	Description: "Customize the behavior of the terminal",
	Icon:        "LuSquareTerminal",
	Settings: map[string]settings.Setting{
		"defaultShell": {
			ID:          "defaultShell",
			Type:        settings.Text,
			Label:       "Default Shell",
			Default:     "/bin/zsh",
			Description: "The default shell to use when opening a terminal",
			Validator: func(value interface{}) error {
				val, ok := value.(string)
				if !ok {
					return settings.ErrSettingTypeMismatch
				}
				path, err := exec.LookPath(val)
				if err != nil {
					return fmt.Errorf("could not find executable: %w", err)
				}
				// make sure the shell exists on the users system
				if _, err := os.Stat(path); os.IsNotExist(err) {
					return fmt.Errorf("could not find shell '%s'", val)
				}

				return nil
			},
		},

		"theme": {
			ID:          "theme",
			Type:        settings.Text,
			Label:       "Theme",
			Default:     "default",
			Description: "The color theme for the terminal",
			Options: []settings.SettingOption{
				{Value: "default", Label: "Default"},
				{Value: "dracula", Label: "Dracula"},
				{Value: "solarized-dark", Label: "Solarized Dark"},
				{Value: "solarized-light", Label: "Solarized Light"},
				{Value: "one-dark-pro", Label: "One Dark Pro"},
				{Value: "monokai", Label: "Monokai"},
				{Value: "nord", Label: "Nord"},
				{Value: "gruvbox-dark", Label: "Gruvbox Dark"},
				{Value: "tokyo-night", Label: "Tokyo Night"},
				{Value: "catppuccin-mocha", Label: "Catppuccin Mocha"},
				{Value: "github-dark", Label: "GitHub Dark"},
				{Value: "github-light", Label: "GitHub Light"},
				{Value: "material-dark", Label: "Material Dark"},
				{Value: "cobalt2", Label: "Cobalt2"},
				{Value: "ayu-dark", Label: "Ayu Dark"},
				{Value: "synthwave-84", Label: "Synthwave '84"},
				{Value: "rose-pine", Label: "Rose Pine"},
				{Value: "kanagawa", Label: "Kanagawa"},
				{Value: "everforest-dark", Label: "Everforest Dark"},
				{Value: "one-light", Label: "One Light"},
			},
		},
		"fontSize": {
			ID:          "fontSize",
			Type:        settings.Integer,
			Label:       "Font Size",
			Default:     12, //nolint:gomnd // this is a reasonable default
			Description: "The size of the font in the terminal",
		},
		"cursorStyle": {
			ID:          "cursorStyle",
			Type:        settings.Text,
			Label:       "Cursor Style",
			Default:     "block",
			Description: "The style of the cursor in the terminal",
			Options: []settings.SettingOption{
				{Value: "block", Label: "Block"},
				{Value: "underline", Label: "Underline"},
				{Value: "bar", Label: "Bar"},
			},
		},
		"cursorBlink": {
			ID:          "cursorBlink",
			Type:        settings.Toggle,
			Label:       "Cursor Blink",
			Default:     true,
			Description: "Whether the cursor should blink in the terminal",
		},
	},
}
